Add RSA xtest cases for pkcs11 in regression 4115

EDVTAZ opened the pull request:
Signed-off by: Gabor Szekely <szvgabor@gmail.com>

EDVTAZ committed:
Add RSA xtest cases for pkcs11 in regression 4115

Signed-off by: Gabor Szekely <szvgabor@gmail.com>

EDVTAZ commented:
Hello @etienne-lms ,

I put together a test (``regression_4115``) based on the ``regression_4006`` for testing RSA with pkcs11. I will be working to make these tests pass. No need to merge this right now since the actual implementation is not done yet, I just wanted to show this to you and ask if you think think it is good.

Thanks,
Gabor

etienne-lms commented on host/xtest/regression_4000.c:
```diff
@@ -2792,11 +2792,11 @@ void run_xtest_tee_test_4110(ADBG_Case_t *c, CK_SLOT_ID slot);
-						sizeof(CK_OBJECT_CLASS) }, \
-		{ CKA_VALUE, (void *)_key_array, sizeof(_key_array) }, \
-	}
+                    sizeof(CK_KEY_TYPE) },  \
```
minor: discard these changes. prefer tabulation for source file consistency.

etienne-lms commented on host/xtest/regression_4000.c:
```diff
@@ -4990,6 +4990,395 @@ static void xtest_tee_test_4006(ADBG_Case_t *c)
+#ifdef CFG_SECURE_KEY_SERVICES
+/*
+ * The test below belongs to the regression 41xx test. As it rely on test
+ * vectors define for the 40xx test, this test sequence in implemented here.
```
minor: s/define/defined/

etienne-lms commented on host/xtest/regression_4000.c:
```diff
@@ -4990,6 +4990,395 @@ static void xtest_tee_test_4006(ADBG_Case_t *c)
+		{ CKA_CLASS, &(CK_OBJECT_CLASS){CKO_PUBLIC_KEY},	
+						sizeof(CK_OBJECT_CLASS) }, 
+		{ CKA_KEY_TYPE,	&(CK_KEY_TYPE){CKK_RSA}, sizeof(CK_KEY_TYPE) }, 
+        { CKA_LABEL, label_asym_pub, sizeof(label_asym_pub)-1 },
```
minor: tab rather than spaces
minor: add spaces around the minus sign (coding style consistency)

etienne-lms commented on host/xtest/regression_4000.c:
```diff
@@ -4990,6 +4990,395 @@ static void xtest_tee_test_4006(ADBG_Case_t *c)
+/*
+ * The test below belongs to the regression 41xx test. As it rely on test
+ * vectors define for the 40xx test, this test sequence in implemented here.
+ * The test below check compliance of crypto algorithms called throug the SKS
```
minor:

s/check/checks/
s/throug/through/
I think you can drop the "SKS" reference, `called throug the PKCS#11 interface.` is fine.

etienne-lms commented on host/xtest/regression_4000.c:
```diff
@@ -4990,6 +4990,395 @@ static void xtest_tee_test_4006(ADBG_Case_t *c)
+		{ CKA_CLASS, &(CK_OBJECT_CLASS){CKO_PRIVATE_KEY},	
+						sizeof(CK_OBJECT_CLASS) }, 
+		{ CKA_KEY_TYPE,	&(CK_KEY_TYPE){CKK_RSA}, sizeof(CK_KEY_TYPE) }, 
+        { CKA_LABEL, label_asym_priv, sizeof(label_asym_priv)-1 },
```
minor: tabulations rather than space chars. Same few lines below.

etienne-lms commented on host/xtest/regression_4100.c:
```diff
@@ -1122,6 +1122,22 @@ static void xtest_tee_test_4109(ADBG_Case_t *c)
+
+	rv = init_lib_and_find_token_slot(&slot);
+	if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK))
+		return;
```
Prefer calling `close_lib()` even if test fails. This is safer if other tests are executed after this one.
Also better move `xtest_tee_test_4115()` implementation after `xtest_tee_test_4114()` implementation. As more and more test are implemented, it is better to keep some consistency.

EDVTAZ replied:
I created ``xtest_tee_test_4115()`` based on the 4110-4112 regressions, and assumed that the ``init_lib_and_find_token_slot()`` function handles it if something fails, as the other regressions are structured the same way.

I placed ``xtest_tee_test_4115()`` here, because it is implemented in the ``regression_4000.c`` file just like the other three next to it, but I can move it if you think it is better that way.

etienne-lms replied:
Got it, my comment was inaccurate. i was thinking about `run_xtest_tee_test_4115(c, slot);`, which indeed already preludes `close_lib()` call whatever its status. My apologizes, i should have read back my comment before posting.

etienne-lms commented on host/xtest/regression_4100.c:
```diff
@@ -1708,3 +1724,5 @@ ADBG_CASE_DEFINE(regression, 4113, xtest_tee_test_4113, /*  TODO: rename 4110 */
 ADBG_CASE_DEFINE(regression, 4114, xtest_tee_test_4114,
 		"Object searches");
+ADBG_CASE_DEFINE(regression, 4115, xtest_tee_test_4115,
+		"run asym crypto tests");
```
minor: s/run asym crypto tests/Compliance of asymmetric cryptography operation/

etienne-lms reviewed (commented):
Hi Gabor,
Thanks a lot for sharing your work. Looks promising. I need to take a bit a time to give useful feedback.
I've pointed some minor comments: use of tabulation for indentation, few typos. I i came into the guts of your tests... i need a bit more time. I'll get back to you.

EDVTAZ committed:
formatting and corrections from PR

etienne-lms commented on host/xtest/regression_4000.c:
```diff
@@ -1730,9 +1730,9 @@ static void xtest_tee_test_4002(ADBG_Case_t *c)
- * The test below check compliance of crypto algorithms called throug the SKS
+ * The test below belongs to the regression 41xx test. As it relies on test
+ * vectors defined for the 40xx test, this test sequence in implemented here.
+ * The test below checks compliance of crypto algorithms called through the
```
:) looks like i should have read my commetns first.
Thanks for the fixes.

etienne-lms commented on host/xtest/regression_4000.c:
```diff
@@ -4990,6 +4990,395 @@ static void xtest_tee_test_4006(ADBG_Case_t *c)
+} MECHANISM_CONVERTER;
+
+// https://crypto.stackexchange.com/questions/37726/in-pkcs1-v2-2-should-mgf-be-based-on-the-same-hash-function-that-rsa-oaep-encr
+// assume optee uses the same hash for mgf, could look it up in source later
```
OP-TEE core currently relies on libtomcrypt which do use the same hash algo for MGF and 
For example: [pkcs_1_oaep_decode.c line 133](https://github.com/OP-TEE/optee_os/blob/3.1.0/core/lib/libtomcrypt/src/pk/pkcs1/pkcs_1_oaep_decode.c#L133) and [pkcs_1_oaep_decode.c line 157](https://github.com/OP-TEE/optee_os/blob/3.1.0/core/lib/libtomcrypt/src/pk/pkcs1/pkcs_1_oaep_decode.c#L157). If another lib is used, this should be checked, but i think it is OK for this test which expect to reflect the existing regression 4006.

etienne-lms commented on host/xtest/regression_4000.c:
```diff
@@ -4990,6 +4990,395 @@ static void xtest_tee_test_4006(ADBG_Case_t *c)
+};
+
+MECHANISM_CONVERTER mechanism_converter[] = {
+    { (CK_MECHANISM_TYPE)CKM_RSA_X_509, (CK_VOID_PTR)NULL, (CK_ULONG)0, (uint32_t)TEE_ALG_RSA_NOPAD, (uint32_t)0 }, //TODO is this correct? https://crypto.stackexchange.com/questions/9593/ckm-rsa-pkcs-vs-ckm-rsa-x-509-mechanisms-in-pkcs11 
```
Look ok to me, as far as i can tell.

etienne-lms commented on host/xtest/regression_4000.c:
```diff
@@ -4990,6 +4990,395 @@ static void xtest_tee_test_4006(ADBG_Case_t *c)
+    { (CK_MECHANISM_TYPE)CKM_SHA384_RSA_PKCS_PSS, (CK_VOID_PTR) &pss_sha384_params, (CK_ULONG)sizeof(pss_sha384_params), (uint32_t)TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA384, (uint32_t)0 },
+    { (CK_MECHANISM_TYPE)CKM_SHA512_RSA_PKCS_PSS, (CK_VOID_PTR) &pss_sha512_params, (CK_ULONG)sizeof(pss_sha512_params), (uint32_t)TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA512, (uint32_t)0 },
+
+    { (CK_MECHANISM_TYPE)CKM_RSA_PKCS, (CK_VOID_PTR)NULL, (CK_ULONG)0, (uint32_t)TEE_ALG_RSAES_PKCS1_V1_5, (uint32_t)0 },   // source: https://cryptosense.com/blog/algorithm-choice-in-pkcs11-part-1-rsa-in-v2-20/
```
approved.

etienne-lms commented on host/xtest/regression_4000.c:
```diff
@@ -4990,6 +4990,395 @@ static void xtest_tee_test_4006(ADBG_Case_t *c)
+
+    { (CK_MECHANISM_TYPE)CKM_RSA_PKCS, (CK_VOID_PTR)NULL, (CK_ULONG)0, (uint32_t)TEE_ALG_RSAES_PKCS1_V1_5, (uint32_t)0 },   // source: https://cryptosense.com/blog/algorithm-choice-in-pkcs11-part-1-rsa-in-v2-20/
+
+    { (CK_MECHANISM_TYPE)CKM_RSA_PKCS_OAEP, (CK_VOID_PTR) &oaep_sha1_params, (CK_ULONG)sizeof(oaep_sha1_params), (uint32_t)TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA1, (uint32_t)0 },   // hash algo in CK_PKCS_OAEP_PARAMS
```
Replace `CK_PKCS_OAEP_PARAMS` with `CK_RSA_PKCS_OAEP_PARAMS` or even can remove comment.

etienne-lms committed:
[fix] Add RSA xtest cases for pkcs11 in regression 4115

Comply to 80 char/line when possible as this is the way the
rest of regression_XXXX.c source files are implemented. This give some
constraints when trying to setup readable code.
Therefore propose few macros routines to easy readability.

Also replace hard coded index in rsa_key_pub_attr and rsa_key_priv_attr
with a function scanning attribute ID.

I've removed or modified some comments but feel free to restore
them if you prefer.

I would prefer initialized structures to explicitly specify the
initialized fields. I saw that I should update my AES flavors params
structures...  Note that I don't think it's useful for CK_ATTRIBUTE
tables so I will keep them as they are.

Last, insure good balancing of calls to Do_ADBG_BeginSubCase() and
Do_ADBG_EncSubCase().

These changes have not been tested but at least should build ok.

Feel free to squashed these changes into your commit proposal
"Add RSA xtest cases for pkcs11 in regression 4115" without
applying my signed-of-by, this is only code style matter.

etienne-lms reviewed (commented):
Regarding the test content and how it reuses regression 4006, I think your code is fine. I've checked the relationship between the CK algo IDs and the OP-TEE (GPD TEE) ones. I think they are ok. See my comments. 

There are still some style issues for this test to be consistent with the other **regression_xxxx.c** files.

Better an example than many comments: please have a look at https://github.com/EDVTAZ/optee_test/pull/1.

EDVTAZ committed:
Merge pull request #1 from etienne-lms/for-gabor

[fix] Add RSA xtest cases for pkcs11 in regression 4115

EDVTAZ committed:
fixed some pointer handling errors

etienne-lms commented:
These changes were merged in branch sks as regression test 4117 later moved as regression test 4217.
